leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/endpoint/ServerCoapMessageTranslator.java

```diff
@@ -67,10 +70,39 @@ public Request createCoapRequest(ClientProfile clientProfile, DownlinkRequest<?
     public <T extends LwM2mResponse> T createLwM2mResponse(ClientProfile clientProfile, DownlinkRequest<T> lwm2mRequest,
             Request coapRequest, Response coapResponse, ServerEndpointToolbox toolbox) {
 
+        ContentFormat contentFormat = coapResponse.getOptions().hasContentFormat()
+                ? ContentFormat.fromCode(coapResponse.getOptions().getContentFormat())
+                : null;
+        if (contentFormat == null && lwm2mRequest instanceof ReadRequest) {
+            contentFormat = ((ReadRequest) lwm2mRequest).getContentFormat();
+        }
+
+        TimestampedLwM2mNode timestampedNode = null;
+        try {
+            List<TimestampedLwM2mNode> timestampedNodes = toolbox.getDecoder().decodeTimestampedData(
+                    coapResponse.getPayload(), contentFormat, new LwM2mPath(1, 0, 1), clientProfile.getModel());
+            timestampedNode = timestampedNodes != null ? timestampedNodes.get(0) : null;
+        } catch (CodecException e) {
+            LOG.debug(String.format("Unable to decode response payload of request [%s]", coapRequest));
+        }
+
         LwM2mResponseBuilder<T> builder = new LwM2mResponseBuilder<T>(coapRequest, coapResponse,
                 clientProfile.getEndpoint(), clientProfile.getModel(), toolbox.getDecoder(), toolbox.getLinkParser());
         lwm2mRequest.accept(builder);
-        return builder.getResponse();
+
+        if (builder.getResponse() instanceof ReadResponse) {
+            if (timestampedNode != null && timestampedNode.isTimestamped()) {
+                ReadResponse readResponse = new ReadResponse(ResponseCode.CONTENT, timestampedNode.getNode(),
+                        timestampedNode, null, coapResponse);
+                return (T) readResponse;
+
+            } else {
+                return builder.getResponse();
+            }
+        } else {
+            return builder.getResponse();
+        }
+
```

Comment:
Are you sure this code is still needed ? this sounds to be done in `LwM2mResponseBuilder`, or I missed something  ?